localstack/aws/handlers/fallback.py

```diff
@@ -0,0 +1,65 @@
+"""Handlers for fallback logic, e.g., populating empty requests or defauling with default exceptions."""
+import logging
+
+from werkzeug.datastructures import Headers
+
+from localstack.http import Response
+
+from ..api import RequestContext
+from ..chain import ExceptionHandler, Handler, HandlerChain
+
+LOG = logging.getLogger(__name__)
+
+
+class InternalFailureHandler(ExceptionHandler):
+    """
+    Exception handler that returns a generic error message if there was an exception and there is no response set yet.
+    """
+
+    def __call__(
+        self,
+        chain: HandlerChain,
+        exception: Exception,
+        context: RequestContext,
+        response: Response,
+    ):
+        if response.response:
```

Comment:
The `response` attribute of `Response` is the underlying container for `data`. `response` is always an iterable of bytes (or strings that are later encoded), whereas `data` is a dynamic property that is the rendered output of `response`.

that means the following things:
* `response.response` can never be None. For an empty response, it will be an empty list. `response.response is not None` will therefore always return true
* in the case that `response.response` contains a generator (in the kinesis streaming case for example) calling `response.data` will block until the entire generator has been consumed, so calling `if request.data` is in general not safe.

for setting data, it means the following:
`data` has getters and setters which operate on strings and bytes. calling `response.data = "foobar"` is the same as calling `response.set_data("foobar")`, which will set `self.response = ["foobar".encode('utf-8')]` (simplified). and also update the Content-Length headers appropriately. so whenever we know that something is a string or bytes, we can safely set `response.data = ...`, whereas if we're not sure whether something may be an iterator, it's better to use `response.set_response(...)` because that will do that check for us (at the cost of doing the check), which will again fall back to `response.data = ...` if the input is a string or bytes.

> Could there be a case here where response.response is falsy, but response.data has already been set (and should be returned)?

not that i can think of. if you set `response.data = b""` (to indicate a deliberately empty response), then `response.response` will be `[b""]` which will be truthy (which is btw exactly the case we are guarding against).
unless you set `response.response` to something very weird that overwrites `__bool__` then yes i guess you could run into that problem. but as you said that doesn't seem very likely to me.

so in summary:
* `response.response` may be set to an iterator -> `response.response` will be truthy
* we cant do `if response.data` checks because it would potentially consume the underlying chunked response stream
* an empty response (`data = ""` which is equivalent to `set_response("")`) -> `response` will be automatically truthy
* `data` never set -> `response` will be falsy

Reply:
Makes perfect sense (should have taken a more careful look at the `data` property). Thanks for the detailed explanation! 👍